### ISSUE TITLE
Create recurring tasks after finishing the original one

### ISSUE THREAD

**Issue by rzelnik** [ransome1/sleek#147]
When I create recurring tasks, the application creates it's next recurrence immediately after saving the first task. Then when I edit the original task (for example change it's priority, etc.), it creates new duplicate recurrences of the original task on each edit. 
When I miss completion of the task on time, it also creates duplicates for each day of delay.

I would suggest to change the behavior to create a next recurrence only after confirmation of completion of the original task ("Mark as done"). This would prevent creation of duplicates.

**Comment by ransome1** [comment]
@rzelnik, I assume you set the due date to today to provoke the automated creation of the recurring task, right? Because just creating a task with recurrence and no due date should not create any new task by automatically.

The big issue here is that sleek (being sleek) doesn't use a database, it rereads the todo.txt file all the time, which so far works out surprisingly well. The issue you're having with changing the first todo, which then doesn't change the already existent recurring todo, is directly connected to this. Due to the lack of a database, both todos are not connected to each other (by an ID for instance) so sleek is having difficulties to change the recurring todo, as it doesn't know that it exists. We could only solve this by either introducing a database or more logic which adds too much complexity to such a simple application.

To be perfectly honest I never used this feature myself, as I don't have recurring tasks and so I'm having a hard time to understand what users do expect from a recurring todo feature. In my opinion your suggestion of cutting the automated creation would remove a lot of unwanted (at on my side) complexity from the code. 

But I am not sure if other users would agree with your suggestion. 

@zerodat, @velardevines, @jason-mehmel, @wsdfhjxc, as far as I remeber you guys are actively working with recurrences, correct? Would you mind sharing your opinion on this? 

The idea is to trigger the creation of a new todo (based on an old todo with recurrence) only once the old todo is being set to complete by hand. This would remove the automated creation of new todos. Currently sleek is automatically creating a new todo once the due date has been reached.

**Comment by rzelnik** [comment]
Yes, I confirm that the automated creation of the recurring task is related to tasks with due date. And yes, I understand that processing the data stored in todo.txt file is somehow difficult compared with working with the database. On the other side, todo.txt-compatibility is a big advantage of this application, despite of some difficulties with processing the data. I would prefer to keep the app simple and todo.txt-based rather than adding a new layer of complexity by synchronizing between the database and the todo.txt file. Maybe the best solution would be to let the user choose in the preferences between automated creation each day/week/etc. and creation after finishing the previous recurrence.

**Comment by ransome1** [comment]
My goal though would be to find a solution that offers a good user experience that comes without a new setting. Especially for a function that needs a lot of explaining. Plus from the telemetry data some users are kindly providing I can see that less than 1% of sleeks users are working with recurrences. Which makes it even harder to justify a dedicated setting for this.

The more I think about it the more I want to deactivate the automated creation. But let's see what the others are saying about it, before we just remove it ;)

**Comment by velardevines** [comment]
I admit to not understanding how Sleek implements the recurrence feature. Nor do I really need to understand how it works.  :-)  

Maintaining compatibility of todo.txt format and keeping Sleek code uncomplicated and easy to maintain are important. 

For the recurrence feature, what is important to me is that a new task is created when I "mark as done" a task with a recurrence value.  

So, I agree with the @rzelnik comment in his first post: "I would suggest to change the behavior to create a next recurrence only after confirmation of completion of the original task ("Mark as done"). This would prevent creation of duplicates." 

Hope this helps.

**Comment by zerodat** [comment]
@ransome1 @rzelnik @velardevines ... yes, recurrences are one of the trickiest parts of the todo logic, but essential for folks like myself who depend on them to keep from forgetting all of the bills, tax returns, and other periodic obligations that we don't want cluttering up our brains.  I have implemented some new features in my own build of sleek, because my workflow depends on having recurrences work like they do in other todo.txt tools I use ([simpletask for android](https://github.com/mpcjanssen/simpletask-android), [topydo](https://github.com/topydo/topydo) cli).    So here is the explanation of what I have working, and I'm curious to hear whether you want to get these features to be widely available in sleek.

1. I have added a new user preference called `generatePastRecurrences` which is on by default to keep sleek working as it does right now: that is, if you read a task with due date say 6 months in the past and the recurrence says `rec:1m`, sleek will immediately generate all of the 6 instances for the past 6 months.  For me, this isn't the desired behavior, because it fills up the file with many instances of the same task.  Instead, I set my user preferences to turn `generatePastRecurrences` to false and then sleek will only create one new recurrence item each time you mark the task completed.  In the example I mentioned, when you mark the 6 months past due task as completed, it will make a new one that is 5 months past due.  So you can step forward one recurrence at a time.  For instance if the task was `pay this month's rent due:2021-01-01 rec:1m` then when you pay the January rent, it will make one only for the February rent, and not generate all of the intermediate ones immediately.  I hope that is a clear explanation! 
2.  Now you might be saying, what if I want to always start the recurrence from today's date rather than from the previous due date?  This would make sense if your task is, for example, `wash the windows due:2021-01-01 rec:1m`.  In this case if you finally got around to washing the windows right now, 6 months late, it would still make sense to just do it again in a month, not do each of the five washings that you missed!  This is sometimes called `non-strict` recurrence, whereas the one where you have to do all five (like paying the rent) is called `strict` recurrence.  As far as I can tell, sleek always uses `strict` recurrence right now, but I have added a new user preference `alwaysStrictRecurrences` which defaults to true for the existing behavior, but I set it to false.  Then the syntax `rec:1m` will give the `non-strict` recurrence which is relative to today's date, for the window washing case.
3. And then I have implemented the syntax `rec:+1m` with the optional plus sign to indicate that I want `strict` recurrence.  So this gives you the option of recurring a month after the previous due date (for the rent case) and also you can skip the plus sign to recur a month after today's (completion) date, for the window-washing case.  This syntax is compatible with simpletask and topydo recurrence behavior.  Unfortunately it can't be used yet because it depends on a bug in jsTodoTxt getting fixed.  I have fixed that bug in my fork of jsTodoTxt and made a pull request several days ago, but haven't heard from the maintainer yet.

Sorry for the long explanation.  I hope that is clear.  For another explanation of recurrences, see also [topydo — recurrence](https://cdn.rawgit.com/bram85/topydo/master/docs/index.html#Recurrence).

Let me know if you have questions.  I could push my changes up to my fork of sleek on github in case you want to take a look, but we really can't fully use them until the upstream bug [Pull Request #30 · jmhobbs/jsTodoTxt](https://github.com/jmhobbs/jsTodoTxt/pull/30) with plus signs is fixed.

-- zerodat

**Comment by rzelnik** [comment]
@zerodat thank you for your explanation. Yes, "washing the windows" vs "paying the rent" are good examples of non-strict vs strict recurrences. Good to see that there are other users with the same need to differentiate between these two types of recurrence. Now I see that both types make sense in different use cases.

**Comment by ransome1** [comment]
@zerodat, if it is not too much effort for you, please feel free to send me a PR and I'll check it. But I need to be honest here, it won't have the highest priority as it is not a critical fix or feature right now.

I think for now I will remove the automated creation as it seems to be connected to more confusion then actual benefits.

**Comment by zerodat** [comment]
@ransome1, yes you could remove the old behavior or change the default to the new behavior.

My PR is a two-line change (one line in each of two files).  To make it default to the new behavior, just change the default value in main.js.

If you would rather remove it completely, then you could just delete the block in todos.mjs instead of checking for the user preference.  No change needed to main.js.  That was what I did originally last week, and it worked well for me, but I wasn't sure whether other users preferred the existing behavior, so I added the user preference.

**Comment by ransome1** [comment]
@rzelnik, auto creation has been removed in https://github.com/ransome1/sleek/releases/tag/v1.0.5

Could you please switch to this version, check it and close this issue, if it is fixed?

**Comment by rzelnik** [comment]
I confirm that the automated creation is inactive as expected, but there is an issue with marking the tasks as done: when I create a task with due date today, recurring daily and mark it as done, sometimes it works well, sometimes it creates a task with different content. When I mark tomorrow's recurrence as done, nothing happens, or it creates a new task with the same due date (tomorrow). Sometimes it is impossible do delete an existing task, etc.

**Comment by ransome1** [comment]
I'll look into it

**Comment by ransome1** [comment]
@rzelnik could you please do me a favor and describe this issue with examples of what you do, what is being created and what actually happens?

I can't find any bug but I have the feeling that you might have a different expectation on how the recurence function is suppose to work than it is actually been implemented.

As for you tomorrow example: If you complete a todo with a recurrence (rec:d), the recurrence time frame of 1 day will be added to the date of completion (which is now). So the first todo will create a new todo with due date set to tomorrow. And if you complete this one now, sleek will again create a todo with rec:d (one day) and add this time frame to the date of completion (now).

I tried to describe how this works in the Wiki: https://github.com/ransome1/sleek/wiki/Recurring-todos/_edit

**Comment by rzelnik** [comment]
I created some testing tasks in a new todo.txt file and tried to mark them done and back, but I can't reproduce any of the previously described issues. I think we can consider it to be working well. If some new issue arises, I will create a separate ticket with a step-by-step sequence how to reproduce the issue.

**Comment by ransome1** [comment]
Thanks @rzelnik !
